File: node/arbitrage.ts
import _ from "lodash";
import BigNumber from "bn.js";
import axios from "axios";
import { GasEstimator } from "./secured-finance.js";

export enum PositionType {
  BORROW = 0,
  LEND = 1,
}

export interface Token {
  name: string; // bytes32
}

export interface Order {
  token: Token;
  price: BigNumber;
  // Maturity date is uint256 in the contract, but we can use BigNumberish to represent it
  maturity: BigNumber;
  posType: PositionType;
  // Amount of tokens in wei
  amount: BigNumber;
}

export interface ArbitrageOpportunity {
  borrowPosition: Order;
  lendPosition: Order;
  profit: BigNumber;
}

type CalcArbitrageParams = {
  borrowGasFee: BigNumber;
  lendGasFee: BigNumber;
  swapFee: BigNumber;
};

type Coin = {
  id: string;
  name: string;
};

const sleep = async (ms: number): Promise<void> => {
  return new Promise((resolve) => setTimeout(() => resolve(), ms));
};

export class ArbitrageEngine {
  private readonly MATURE_PRICE = 10_000;

  private _arbitrageOpportunities: Record<string, ArbitrageOpportunity[]> = {};
  private _debugMode: boolean = false;

  private static readonly SUPPORTED_TOKENS: Array<Coin> = [
    { id: "ETH".toUpperCase(), name: "ETH" },
    { id: "FIL".toUpperCase(), name: "EFIL" },
    { id: "BTC".toUpperCase(), name: "WBTC" },
    { id: "USDC".toUpperCase(), name: "USDC" },
  ];

  private tokenPricesInUsd: Record<string, number>;

  // TODO: Receive a list of tokens to calculate arbitrage opportunities as input
  constructor(private readonly gasEstimator: GasEstimator, isDebug: boolean = false) {
    this._debugMode = isDebug;

    if (this._debugMode) {
      this.tokenPricesInUsd = {
        ETH: 1,
        EFIL: 1,
        WBTC: 1,
        USDC: 1,
      };
    } else {
      this.fetchTokenPrices(ArbitrageEngine.SUPPORTED_TOKENS);
    }
  }

  protected parseInput(): Array<Order> {
    return [];
  }

  public get arbitrageOpportunities(): Record<string, ArbitrageOpportunity[]> {
    return this._arbitrageOpportunities;
  }

  private async fetchTokenPrices(coins: Coin[]): Promise<void> {
    while (true) {
      for (const coin of coins) {
        const { data } = await axios.get(
          `https://api.binance.com/api/v3/ticker/price/?symbol=${coin.id}USDT`
        );
        console.log(`Fetched price for ${coin.name}: ${data.price}`);
        this.tokenPricesInUsd[coin.name] = parseFloat(data.price);
      }
      await sleep(1000 * 60 * 5);
    }
  }

  /**
   * Creates order
   */
  private createOrder() {}


  private yearToFraction(maturity: number): number {
    const oneDayInSeconds = 24 * 60 * 60; // number of seconds in one day
    const nowInSeconds = Math.floor(Date.now() / 1000); // current time in seconds
    const secondsToMaturity = maturity - nowInSeconds; // number of seconds between now and maturity
    const daysToMaturity = Math.ceil(secondsToMaturity / oneDayInSeconds); // round up to get number of days

    return daysToMaturity / 365;
  }

  /**
   * Converts a loan price to a rate
   * @returns rate in basis points
   */
  private calculateRate(price: number, maturity: number) {
    return (this.MATURE_PRICE - price) / (price * this.yearToFraction(maturity));
  }

  /**
   * Calculates arbitrage opportunities for a single maturity date
   * @param positions
   * @returns Array of arbitrage opportunities
   */
  private _calculateArbitrageOpportunity(
    positions: Order[],
    params: CalcArbitrageParams = {
      borrowGasFee: new BigNumber(0),
      lendGasFee: new BigNumber(0),
      swapFee: new BigNumber(0),
    }
  ): Array<ArbitrageOpportunity> {
    const arbitrageOpportunities: Array<ArbitrageOpportunity> = [];

    const [borrowPositions, lendPositions] = _.partition(
      positions,
      (pos) => pos.posType === PositionType.BORROW
    );

    // Iterate through every borrow position
    borrowPositions.forEach((borrowPosition) => {
      // Iterate through every lend position
      lendPositions.forEach((lendPosition) => {
        // Skip if the tokens are the same
        if (borrowPosition.token.name === lendPosition.token.name) {
          return;
        }

        // In crypto "unit" is the smallest denomination of a token (1/100)
        const borrowPositionRate = this.calculateRate(borrowPosition.price.toNumber(), borrowPosition.maturity.toNumber());
        const lendPositionRate = this.calculateRate(lendPosition.price.toNumber(), lendPosition.maturity.toNumber());

        const rateDifferential = lendPositionRate - borrowPositionRate;

        if (rateDifferential < 0) {
          return;
        }

        const carryTradeAmount = rateDifferential * borrowPosition.amount.toNumber();
        const carryTradeAmountInUsd = carryTradeAmount * this.tokenPricesInUsd[borrowPosition.token.name];
        
        // Depends on gas price, but we can assume it's 1 gwei
        const dexSwapFee = params.swapFee;

        // Calculate based on "borrow" + "lend" amounts
        const borrowGasFees = params.borrowGasFee;
        const lendGasFees = params.lendGasFee;
        
        const profit = new BigNumber(carryTradeAmountInUsd).sub(dexSwapFee).sub(borrowGasFees).sub(lendGasFees);

        // If we can borrow the token at a lower price than we can lend it, we have an arbitrage opportunity
        if (profit.gt(new BigNumber(0))) {
          arbitrageOpportunities.push({
            borrowPosition,
            lendPosition,
            profit,
          });
        }
      });
    });

    // Sort by profit (descending order)
    arbitrageOpportunities.sort((a, b) => b.profit.cmp(a.profit));

    return arbitrageOpportunities;
  }

  /**
   * Calculates arbitrage opportunities for each maturity date
   * Takes gas fees into account
   * @param positions Array of positions
   * @returns Record of maturity date to array of arbitrage opportunities
   */
  public async calculateArbitrageOpportunities(positions: Order[]): Promise<void> {
    const positionsByMaturity = _.groupBy(positions, (pos) =>
      pos.maturity.toString()
    );

    // Calculate gas fees
    const [lendGasFee, borrowGasFee, swapFee] = await Promise.all([
      this.gasEstimator.estimateLendingFee(),
      this.gasEstimator.estimateBorrowingFee(),
      this.gasEstimator.estimateSwapFees(),
    ])
    console.log(`Lend gas fee: ${lendGasFee.toString()}`)
    console.log(`Borrow gas fee: ${borrowGasFee.toString()}`)
    console.log(`Swap fee: ${swapFee.toString()}`)
    
    const gasFeeParams = this._debugMode ? {
      borrowGasFee: new BigNumber(lendGasFee.toString()),
      lendGasFee: new BigNumber(borrowGasFee.toString()),
      swapFee: new BigNumber(swapFee.toString()),
    } : undefined

    // Iterate through every maturity date
    Object.entries(positionsByMaturity).forEach(
      ([maturity, sameMaturityPositions]) => {
        this.arbitrageOpportunities[maturity] =
          this._calculateArbitrageOpportunity(sameMaturityPositions);
      }
    );
  }

  public execute() {
    // "depositAndCreatorOrder"
    // "createOrder"
  }
}

File: rust/src/arbitrage.rs
use ethers::types::U256;
use num_enum::TryFromPrimitive;
use std::{collections::HashMap};
use std::time::{SystemTime, UNIX_EPOCH};

use crate::utils;

#[derive(Debug, PartialEq, Clone, TryFromPrimitive)]
#[repr(u32)]
pub enum PositionType {
  LEND = 0,
  BORROW = 1,
}

#[derive(Clone, Debug)]
pub struct Token {
  pub name: Box<str>,
}

#[derive(Clone, Debug)]
pub struct Order {
  pub token: Token,
  // Price form 0 - 10_000
  pub price: i16,
  // Maturity date is uint256 in the contract, but we can use BigNumberish to represent it
  pub maturity: u64,
  pub pos_type: PositionType,
  // Amount of tokens in wei
  pub amount: U256,
}

#[derive(Debug, Clone)]
pub struct Opportunity {
  pub borrow_position: Order,
  pub lend_position: Order,
  // In USD
  pub profit: f64,
}

pub struct Arbitrage {
  pub opportunities: HashMap<i64, Vec<Opportunity>>,
  is_debug_mode: bool,
  token_prices_in_usd: HashMap<String, f64>,
}


// TODO: How to make "private class variables" in Rust
impl Arbitrage {
  #[must_use]
  pub fn new(is_debug_mode: bool) -> Arbitrage {
    let mut arbitrage = Arbitrage {
      is_debug_mode,
      opportunities: HashMap::new(),
      token_prices_in_usd: HashMap::new()
    };

    // TODO: Get prices from API
    arbitrage.token_prices_in_usd.insert("ETH".to_string(), 1.0);
    arbitrage.token_prices_in_usd.insert("EFIL".to_string(), 1.0);
    arbitrage.token_prices_in_usd.insert("WBTC".to_string(), 1.0);
    arbitrage.token_prices_in_usd.insert("USDC".to_string(), 1.0);

    arbitrage
  }

  fn maturity_to_fraction(maturity: u64) -> f64 {
    let one_day_in_seconds: f64 = 24.0 * 60.0 * 60.0;
    let now_in_seconds: u64;

    let now_duration = SystemTime::now().duration_since(UNIX_EPOCH);
    match now_duration {
      Ok(duration) => {
        let seconds_since_epoch = duration.as_secs() as u64;
        now_in_seconds = seconds_since_epoch;
      },
      Err(msg) => panic!("Problem calculating duration since epoch {:?}", msg.to_string())
    }

    let seconds_to_delay = (maturity - now_in_seconds) as f64;
    let days_to_maturity: f64 = seconds_to_delay / one_day_in_seconds;
    
    return days_to_maturity / 365.0;
  }

  fn calculate_rate(price: i32, maturity: u64) -> f64 {
    let mature_price: i32 = 10_000;

    return (mature_price - price) as f64 / ((price as f64) * Arbitrage::maturity_to_fraction(maturity));
  }

  fn calculate_arbitrage<'a>(
    &self,
    orders: &Vec<Order>
  ) -> Vec<Opportunity> {
    let mut opportunities: Vec<Opportunity> = vec![];

    let mut borrow_orders: Vec<&Order> = vec![];
    let mut lend_orders: Vec<&Order> = vec![];

    for order in orders {
      if order.pos_type == PositionType::BORROW {
        borrow_orders.push(order)
      } else {
        lend_orders.push(order);
      }
    }

    // TODO: Fasten naive algorithm with pre-computation, with large datasets
    for b_order in &borrow_orders {
      for l_order in &lend_orders {
        if b_order.token.name == l_order.token.name {
          continue;
        }

        let borrow_order_rate = Arbitrage::calculate_rate(
          b_order.price as i32,
          b_order.maturity
        );
        let lend_order_rate = Arbitrage::calculate_rate(
          l_order.price as i32,
          l_order.maturity
        );

        let rate_differential = lend_order_rate - borrow_order_rate;

        if rate_differential < 0.0 {
          continue;
        }

        let carry_trade_amount: f64 = rate_differential * (b_order.amount.as_u64() as f64);
        let token_name = b_order.token.name.to_string();

        let token_price: f64;
        match self.token_prices_in_usd.get(&token_name) {
          Some(price) =>  token_price = *price,
          None => continue,
        }
        let carry_trade_amount_usd: f64 = carry_trade_amount * token_price;

        // TODO: Pass as params from "ethers.rs", in USD
        let dex_swap_fee = 0;
        let borrow_gas_fee = 0;
        let lend_gas_fee = 0;

        let profit: f64 = carry_trade_amount_usd - (dex_swap_fee + borrow_gas_fee + lend_gas_fee) as f64;

        if profit > 0.0 {
          let opportunity = Opportunity {
            borrow_position: (**b_order).clone(),
            lend_position: (**l_order).clone(),
            profit, 
          };

          opportunities.push(opportunity);
        }
      }

    }

    opportunities
  }

  pub fn calculate_arbitrage_opportunities(
    &mut self,
    positions: &Vec<Order>,
  ) {
    let orders_by_maturity = utils::group_by(positions, | pos | pos.maturity);


    // TODO: Calculate gas fees with "ethers.rs

    for orders in orders_by_maturity.into_iter() {
      let maturity = orders.0;
      let orders_vec_ref = utils::to_ref(orders.1);


      self.opportunities.insert(maturity as i64, self.calculate_arbitrage(&orders_vec_ref));
    }
  }
}

